### ISSUE TITLE
Create cross-walk table from PRMS segment ID to NHDPlusV2 COMID for DRB

### ISSUE THREAD

**Issue by lekoenig** [USGS-R/drb-inland-salinity-ml#21]
Develop cross walk table between PRMS and NHD; check matched NHD reach assignments

**Comment by lekoenig** [comment]
We ultimately want to be able to crosswalk between PRMS segments and NHDPlusV2 flowlines to take advantage of rich attribute datasets that are referenced to NHDPlus. Below I've added some notes that describe what that process looks like right now. 

The geospatial fabric (GF)/PRMS network is less dense relative to the NHD, and PRMS reaches are often aggregated versions of NHD flowline segments. The [GF](https://www.sciencebase.gov/catalog/item/535eda80e4b08e65d60fc834) contains a variable called `POI_ID` that corresponds to the NHDPlus COMID at the downstream end of the PRMS reach. Given this schema, we can use a tabular join between GF and the [DRB-PRMS segments](https://www.sciencebase.gov/catalog/item/5f6a285d82ce38aaa244912e) to identify the NHDPlusV2-COMID at the downstream end of each reach, based on the shared segment ID. 

![tabular_join_example](https://user-images.githubusercontent.com/8785034/145833326-71ad6259-5764-48ef-a139-b0ec17e8fecc.png)

However, in inspecting the results I've noticed approximately 20-25 edge cases where the tabular join may not result in the expected PRMS-NHD match. These edge cases generally fall into 3 categories:  

1. As best as I can tell, the GF flowline segments were referenced to NHDPlusV1, necessitating an additional step to crosswalk between the COMID and `POI_ID` from NHDPlusV1 to the NHDPlusV2-COMID. In most cases, these COMID's are a 1-1 match. There are a couple of cases where the crosswalk between NHDPlusV1 and NHDPlusV2 is not 1-1, and so the COMID referenced in `POI_ID` doesn't correspond to the NHDPlusV2-COMID at the bottom of the reach: 

![xwalk_not_1_1](https://user-images.githubusercontent.com/8785034/145845096-dae52c62-c1a6-40e7-ab71-d10ee9ae2117.png)

2. For the DRB-PRMS network, a subset of segments were split such that  segment start and endpoints were connected by one reach with no intersecting reaches. So in a couple of cases, two DRB-PRMS reaches (indicated by unique `subsegid`'s) share the same NHD reach:  

![two_segs](https://user-images.githubusercontent.com/8785034/145845954-0a597894-f868-403a-aa99-657e6eb51139.png)

3. Finally, there are ~20 cases where the NHD assignment based on the tabular join doesn't _really_ seem like the best possible match:  

![incorrect_match](https://user-images.githubusercontent.com/8785034/145846997-496e3447-8841-4db8-a2be-411008fd4276.png)

The edge cases described above could result in incorrect attribute assignments for our DRB-PRMS reaches, so to handle these, I created a function to perform the spatial join between the DRB-PRMS reaches and NHDPlusV2 flowlines. Simple intersections result in the edge cases described in 3 above, so instead I identify which NHD flowlines have >1 overlapping nodes (suggesting that the NHD flowline overlaps the PRMS reach as opposed to just touching it):  

![spatial_join](https://user-images.githubusercontent.com/8785034/145848916-2294d7e3-8c73-40c6-b077-3f7d9c7d8f21.png)

We can save columns such as `comid_all` and `comid_down` which contain all of the overlapping NHD segments within a PRMS reach and the most downstream NHDPlusV2-COMID, respectively. Either of these variables could be used to link the NHD attribute information to each PRMS reach (e.g. by addition or by diffs between current segment-upstream segment).  The downside is that the spatial join is more time-consuming than the simple tabular join - for example, it takes ~1.5 hr to pair all 459 DRB reaches on my local machine. The edge cases described here represent a relatively small proportion (~5%) of the network, so I'm curious what you think about this potential trade-off between pipeline build-time/complexity and accuracy for this NHD-PRMS crosswalk task. @jds485 @jsadler2 @msleckman

**Comment by jsadler2** [comment]
Interesting. It's great that you've taken such a thorough look at this. I have a couple of questions.
1. You say there are 20-25 edge cases total and then you say there are ~20 cases of `3`. Does that mean almost all of them are of the `3` type? 
2. How did you find these cases? Was it programmatically or did you have to manually inspect them? I guess `2` can be found programmatically but maybe not `1` or `3`?

If you can find them programmatically, maybe we can just do the spatial joins for the "problem" segments and the tabular for the rest. This would theoretically result in the same answer but would only have to do the "long way" for 5% of the segments (20-25 of 459). I guess even if you have to find them manually, you've pretty much done that already, so you could just list the "problem segments" in a vector and write a comment of how you found these.

Another idea for doing the spatial join - could you do a buffer on the PRMS streams to convert them into polygons and then do a `st_intersection` to get how much of each line is in each polygon? That might be faster computationally than converting the lines to points.

**Comment by lekoenig** [comment]
> You say there are 20-25 edge cases total and then you say there are ~20 cases of 3. Does that mean almost all of them are of the 3 type?

Yep, nearly all of the edge cases are type 3 (20 cases versus 1-2 in each of the previous 2 groupings).

> How did you find these cases? Was it programmatically or did you have to manually inspect them? I guess 2 can be found programmatically but maybe not 1 or 3?

I'd say a mix. Groups 1-2 I found programmatically whereas group 3 was based on visual inspection - I had a script save a plot similar to the ones above for each reach, and then quickly flipped through those to find the unexpected matches. 

> Another idea for doing the spatial join - could you do a buffer on the PRMS streams to convert them into polygons and then do a st_intersection to get how much of each line is in each polygon? 

 I did try out `st_intersects` and `st_intersection`, but I like the way that SO post filters the intersecting lines by length - I hadn't added that step even though that's what I'm getting at in searching for overlapping nodes > 1. I just briefly tried this out on 20 segments and the approach you suggest returns the same values for `comid_down` and is _a lot_ faster! Woo! I'll double-check some `comid_all` comparisons but then I think we can apply this new function to the full network - the speed savings should get rid of the need to mix and match tabular and spatial joins, which would make the code more complicated (and less generalizable to other watersheds). Thanks for these thoughts, @jsadler2!

![time_compare](https://user-images.githubusercontent.com/8785034/145874993-bea1d919-7bb6-47e5-93aa-6a0a80741636.png)

![st_intersection](https://user-images.githubusercontent.com/8785034/145875005-bae6d143-2db1-40a3-9aa4-247f03ad9341.png)
